### PR TITLE
[automation] Script file watchers now reusable for language-specific versions

### PR DESCRIPTION
**Pull request by jpg0**
Extracted reusable parts of script watching to allow language-specific instantiations:

- Moved to non-internal package
- Extracted logic into non-osgi-component classes
- Added back in basic osgi-component versions

This change is specifically to prepare for JS-specific scripts to operate as standard NodeJS which is slightly different to the existing openHAB mechanism, as discussed with @kaikreuzer. No functionality change to core with this change, except the creation of the script directory if not already present. PR to use this functionality in openhab-addons to follow. 

Signed-off-by: Jonathan Gilbert <jpg@trillica.com>

### CODE REVIEW

**Comment by jpg0** [comment]
Related PR to actually use this: https://github.com/openhab/openhab-addons/pull/11719

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/internal/loader/DefaultScriptFileWatcher.java
Comment (digitaldan) [inline bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/internal/loader/DefaultScriptFileWatcher.java:41]: Looks like some merge conflicts did not get resolved?

**Comment by digitaldan** [comment]
@cweitkamp and @kaikreuzer we have a number of changes we are trying to get in to provide proper support for the GraalVM script engine.  This PR as well as #2595 will allow us to use both Nashorn and GraalVM, side by side, in both UI rules as well as files.  Right now the two engines kinda clobber each other and are hurting us more then helping when you try and install GraalVM.  This should really help us start to transition to GraalVM as our preferred scripting engine and eventually remove Nashorn in a graceful way over time.

**Comment by digitaldan** [comment]
Any chance of a review or feedback @kaikreuzer @cweitkamp ? Thanks!

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/ScriptFileWatcher.java
Comment (cweitkamp) [inline bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/ScriptFileWatcher.java:60]: Should this be an abstract class now?

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/DependencyTracker.java
Comment (cweitkamp) [inline bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/DependencyTracker.java:34]: Can you help me with this? Will it be necessary for every script automation bundle to implement it's own `DependencyTracker` service after this change?

**Review by cweitkamp — COMMENTED** [review]
Thanks. I support the idea behind this PR. I like to see a review from @kaikreuzer as this has been discussed with him earlier.

I left two questions for my personal understanding.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/ScriptFileWatcher.java
Comment (jpg0) [reply bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/ScriptFileWatcher.java:60]: It could be, although I left it as concrete as technically code could (usefully) instantiate an instance without extending it. Happy to make it abstract if you'd prefer.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/DependencyTracker.java
Comment (jpg0) [reply bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/DependencyTracker.java:34]: It won't be necessary (it's entirely optional), but it will be required _if_ they want explicit dependency tracking - e.g. script reloading if dependencies change (no addon actually does track dependencies except the JS one though afaik). You can see how the `DependencyTracker` is actually added in the associated JS PR.

- https://github.com/openhab/openhab-addons/pull/11719

Comment (cweitkamp) [reply bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/DependencyTracker.java:34]: Thanks for the link. I saw your changes over there. That's why I am asking.

Please correct me if I am wrong. For now the dependency tracking is available by default for everyone automation script bundle. With this PR it will be turned of which will be a great loss if we do not implement it immediately for every existing extension.

Or does it work only for JS-based automation?

Comment (jpg0) [reply bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/DependencyTracker.java:34]: Currently it only works for JS-based automation.

The reason for this is that there are two pieces to be implemented for it to work:
- Monitoring of any library file changes (where this class is involved)
- Establishing the dependency graph of which script depends on which library (so we can determine which scripts to reload if a library file changes)

Whilst _technically_ all languages currently get the first part for free (because currently all libraries are in the same source root), only the JS addon actually implements the second part. This second part can be implemented by any language by notifying the dependency tracker of explicit dependencies, which the JS addon does by hooking into the module loading system and recording them as they happen.

So there is no loss of existing functionality here; only slightly more work (at the benefit of much greater flexibility) for other languages to add dependency tracking at some point.

**Comment by digitaldan** [comment]
@kaikreuzer bump!

**Comment by kaikreuzer** [comment]
@digitaldan: As @cweitkamp already did the review, I'll leave it to him to check if his feedback has been addressed and to merge the PR.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/internal/loader/DefaultScriptFileWatcher.java
Comment (kaikreuzer) [inline bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/internal/loader/DefaultScriptFileWatcher.java:27]: please remove the line break here

**Review by kaikreuzer — CHANGES_REQUESTED** [review]
Thanks, code looks good to me, but I haven't tested it - I assume you did.
Just some minor style comments below and also waiting for @cweitkamp's approval of it.

**Comment by digitaldan** [comment]
Thanks!  Yes i have been testing it for about a week without issues.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/ScriptFileWatcher.java
Comment (cweitkamp) [inline bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/ScriptFileWatcher.java:108]: If  you have to touch it again, please remove this empty line too.

**Review by cweitkamp — APPROVED** [review]
Thanks for the feedback. I am fine with the code.

Two minor style comments left before this can be merged. We can focus on functionality now. Missing nullness annotations can be added later.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/ScriptFileWatcher.java
Comment (cweitkamp) [reply bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/ScriptFileWatcher.java:60]: No, not really necessary.

File: bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/ScriptFileWatcher.java
Comment (jpg0) [reply bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/ScriptFileWatcher.java]: I've removed this. I also thought I'd take a quick look at nullness annotations in case it was a trivial addition, and it was, so I added these too.

**Review by kaikreuzer — APPROVED** [review]

**Comment by jpg0** [comment]
So this builds and tests fine on my machine, is the failing build is due to flakiness in the integration tests? If so should I force re-push with no material changes to force another build? Or can we ignore it?

**Comment by kaikreuzer** [comment]
As the PR build on Jenkins succeeded, I guess we can ignore it.

**Comment by digitaldan** [comment]
@jpg0 https://ci.openhab.org/job/openHAB-Core/880/ is building right now which i think will push updated core bundles to the repos.

**Pull request merged**
